### PR TITLE
Upgrade `checkAnagram` function & Fixes

### PR DESCRIPTION
**Pull request by fahimfaisaal**
# Welcome to the JavaScript community

[Open in Gitpod](https://gitpod.io/from-referrer/)&nbsp;[know more](https://www.gitpod.io/docs/pull-requests/)

### Describe your change:

* [x] Optimize the algorithm via `Array.prototype.reduce` & `String.prototype.replace` methods
* [x] Fix a bug case-sensitive to insensitive by using `RegExp`
* [x] Add new JS Doc
* [x] Add new test cases for invalid input & case-insensitive
* [x] **Fixed:** #901 

### Checklist:
* [x] I have read [CONTRIBUTING.md](https://github.com/TheAlgorithms/Javascript/blob/master/CONTRIBUTING.md).
* [x] This pull request is all my own work -- I have not plagiarized.
* [x] I know that pull requests will not be merged if they fail the automated tests.
* [x] If this pull request resolves one or more open issues then the commit message contains `Fixes: #{$ISSUE_NO}`.

### CODE REVIEW

File: String/CheckAnagram.js [hunk 1]
```diff
@@ -1,47 +1,28 @@
-// An [Anagram](https://en.wikipedia.org/wiki/Anagram) is a string that is a word or phrase formed by rearranging the letters of a different word or phrase, typically using all the original letters exactly once.
-// Anagram check is case sensitive; i.e. Aba and aba is not a anagram.
-// inputs are strings i.e. str1 and str2
+/**
+ * @function checkAnagram
+ * @param {string} str1
+ * @param {string} str2
+ * @returns {boolean}
+ * @description An [Anagram](https://en.wikipedia.org/wiki/Anagram) is a string that is a word or phrase formed by rearranging the letters of a different word or phrase, typically using all the original letters exactly once. Anagram check is case sensitive;
```
Comment (raklaptudirm) [inline String/CheckAnagram.js:6]:
```suggestion
 * @description An [Anagram](https://en.wikipedia.org/wiki/Anagram) is a string that is a word or phrase formed by rearranging the letters of a different word or phrase, typically using all the original letters exactly once. Anagram check is not case sensitive;
```

File: String/CheckAnagram.js [hunk 2]
```diff
@@ -1,47 +1,28 @@
...
-
-  // Now check if all entries in hashmap has zeros.
-
-  for (const key in str1CharCount) {
-    if (str1CharCount[key] !== 0) return false
-  }
-
-  return true
+  // str1 converted to array and traverse the each letter of str1 by reduce method
+  return ![...str1].reduce(
```
Comment (raklaptudirm) [inline String/CheckAnagram.js:22]: This line may be confusing for people who a new to Javascript. I would suggest properly explaining the substitution and boolean type casting steps in the comments.

**Review by raklaptudirm — CHANGES_REQUESTED** [review]

**Comment by kikar** [comment]
> Optimize the algorithm via Array.prototype.reduce & String.prototype.replace methods
Looks like the algorithm is less efficient now. It was O(n) and now is O(n^2)

**Comment by raklaptudirm** [comment]
@kikar I think the new implementation can teach more Javascript concepts than the previous one, which is our primary aim. That said, I think it will be better to have the two implementations separately @fahimfaisaal.

File: String/CheckAnagram.js (hunk 2 above)
Comment (fahimfaisaal) [reply String/CheckAnagram.js:22]: I think so, thanks for your suggestion.

**Comment by fahimfaisaal** [comment]
@kikar to my knowledge, the `String.prototype.replace` method used regex that's why the operation is very cheap, it's not traversing the full string `str2Acc` in every traverse of `Array.prototype.reduce` method. it just removes **1 char** from `str2Acc` in 1 `Array.prototype.reduce` method traverse. which char is just matched to the `cur` variable.

**Comment by fahimfaisaal** [comment]
@kikar to my knowledge, the `String.prototype.replace` method used **regex** that's why the operation is very cheap, it's not traversing the full string `str2Acc` in every traverse of `Array.prototype.reduce` method. it just removes `1 char` from `str2Acc` in 1 `Array.prototype.reduce` method traverse. which char is just matched to the cur variable.

So, the algorithm couldn't be O(n^2)

**Comment by fahimfaisaal** [comment]
@kikar  to my knowledge, the `String.prototype.replace` method used **regex** that's why the operation is very cheap, cause it's not traversing the full string `str2Acc` in every traverse of `Array.prototype.reduce` method. it just removes **1 char** from `str2Acc` in 1 `Array.prototype.reduce` method traverse. which char is just matched to the `cur` variable.

So, the algorithm couldn't be **O(n^2)** complexity.

**Comment by kikar** [comment]
> @kikar to my knowledge, the `String.prototype.replace` method used **regex** that's why the operation is very cheap, cause it's not traversing the full string `str2Acc` in every traverse of `Array.prototype.reduce` method. it just removes **1 char** from `str2Acc` in 1 `Array.prototype.reduce` method traverse. which char is just matched to the `cur` variable.
> 
> So, the algorithm couldn't be **O(n^2)** complexity.

Even if use regex, the code needs to find the first occurrence of the letter selected. Take the case of the reversed string.
For each letter from str1 you need to traverse the whole str2 to find the letter to remove.

**Comment by fahimfaisaal** [comment]
> @kikar I think the new implementation can teach more Javascript concepts than the previous one, which is our primary aim. That said, I think it will be better to have the two implementations separately @fahimfaisaal.

@raklaptudirm I think if I optimize my algorithm, It's shouldn't add two more implementations. please let me know what could I do.

File: String/CheckAnagram.js (hunk 1 above)
Comment (fahimfaisaal) [reply String/CheckAnagram.js]: FIXED

**Comment by raklaptudirm** [comment]
Since the two implementations are very different, I would suggest to keep and test both.

**Comment by fahimfaisaal** [comment]
@raklaptudirm I reckon, if we add this algorithm it's will replace previous algos.
```javascript
const str1List = Array.from(str1.toUpperCase()) // str1 to array 

  // get the occurrences of characters of str1 by using HashMap
  const str1Occurs = str1List.reduce(
    (map, char) => (
      map.set(char, map.get(char) + 1 || 1), map
    ),
    new Map()
  )

  for (const char of str2.toUpperCase()) {
    // if char has not exist to the map it's return false
    if (!str1Occurs.has(char)) {
      return false;
    }

    let getCharCount = str1Occurs.get(char)
    str1Occurs.set(char, --getCharCount)

    getCharCount === 0 && str1Occurs.delete(char)
  }

  return true;
}
```

**Comment by raklaptudirm** [comment]
Both do not need to have the same name.

**Comment by fahimfaisaal** [comment]
Ok I will add these two with different names, but could I add these two algos in One file?

**Comment by raklaptudirm** [comment]
Yes.

**Comment by fahimfaisaal** [comment]
Ok, thanks for your rapid response.

File: String/CheckAnagram.js [hunk 3]
```diff
@@ -18,11 +19,58 @@ const checkAnagram = (str1, str2) => {
...
+/**
+ * @function checkAnagramViaMap
+ * @description - check anagram via using HashMap
+ * @param {string} str1
+ * @param {string} str2
+ * @returns {boolean}
+* @example - checkAnagramViaMap('node', 'deno') => true
+ * @example - checkAnagramViaMap('Eleven plus two', 'Twelve plus one') => true
+ */
+const checkAnagramViaMap = (str1, str2) => {
```
Comment (raklaptudirm) [inline String/CheckAnagram.js:42]:
```suggestion
const checkAnagramMap = (str1, str2) => {
```

**Review by raklaptudirm — CHANGES_REQUESTED** [review]
Rename the functions to remove `via` from the name.

File: String/CheckAnagram.js [hunk 4]
```diff
@@ -1,13 +1,14 @@
...
  * @returns {boolean}
- * @description An [Anagram](https://en.wikipedia.org/wiki/Anagram) is a string that is a word or phrase formed by rearranging the letters of a different word or phrase, typically using all the original letters exactly once. Anagram check is not case-sensitive;
- * @example - checkAnagram('node', 'deno') => true
- * @example - checkAnagram('Eleven plus two', 'Twelve plus one') => true
+ * @description - check anagram with the help of Regex
+ * @example - checkAnagramViaRegex('node', 'deno') => true
+ * @example - checkAnagramViaRegex('Eleven plus two', 'Twelve plus one') => true
  */
-const checkAnagram = (str1, str2) => {
+const checkAnagramViaRegex = (str1, str2) => {
```
[the remark this answers is not in the document]
Comment (fahimfaisaal) [reply String/CheckAnagram.js]: Done

**Comment by fahimfaisaal** [comment]
You're really strict reviewer. 😀

**Review by raklaptudirm — APPROVED** [review]

**Pull request merged**